src/state/farms/fetchFarmsPrices.ts

```diff
@@ -0,0 +1,85 @@
+import BigNumber from 'bignumber.js'
+import { BIG_ZERO } from 'utils/bigNumber'
+import { filterFarmsByQuoteToken } from 'utils/farmsPriceHelpers'
+import { Farm } from 'state/types'
+
+const getFarmFromTokenSymbol = (farms: Farm[], tokenSymbol: string, preferredQuoteTokens?: string[]): Farm => {
+  const farmsWithTokenSymbol = farms.filter((farm) => farm.token.symbol === tokenSymbol)
+  const filteredFarm = filterFarmsByQuoteToken(farmsWithTokenSymbol, preferredQuoteTokens)
+  return filteredFarm
+}
+
+const getFarmBaseTokenPrice = (farm: Farm, quoteTokenFarm: Farm, bnbPriceBusd: BigNumber) => {
+  if (farm.quoteToken.symbol === 'BUSD') {
+    return new BigNumber(farm.tokenPriceVsQuote)
+  }
+
+  if (farm.quoteToken.symbol === 'wBNB') {
+    return bnbPriceBusd.times(farm.tokenPriceVsQuote)
+  }
+
+  // We can only calculate profits without a quoteTokenFarm for BUSD/BNB farms
+  if (!quoteTokenFarm) {
+    return BIG_ZERO
+  }
+
+  // Possible alternative farm quoteTokens:
+  // UST (i.e. MIR-UST), pBTC (i.e. PNT-pBTC), BTCB (i.e. bBADGER-BTCB), ETH (i.e. SUSHI-ETH)
+  // If the farm's quote token isn't BUSD or wBNB, we then use the quote token, of the original farm's quote token
+  // i.e. for farm PNT - pBTC we use the pBTC farm's quote token - BNB, (pBTC - BNB)
+  // from the BNB - pBTC price, we can calculate the PNT - BUSD price
+  if (quoteTokenFarm.quoteToken.symbol === 'wBNB') {
+    const quoteTokenInBusd = bnbPriceBusd.times(quoteTokenFarm.tokenPriceVsQuote)
+    return new BigNumber(farm.tokenPriceVsQuote).times(quoteTokenInBusd)
+  }
+
+  if (quoteTokenFarm.quoteToken.symbol === 'BUSD') {
+    const quoteTokenInBusd = quoteTokenFarm.tokenPriceVsQuote
+    return new BigNumber(farm.tokenPriceVsQuote).times(quoteTokenInBusd)
+  }
+
+  // Catch in case token does not have immediate or once-removed BUSD/wBNB quoteToken
+  return BIG_ZERO
+}
+
+const getFarmQuoteTokenPrice = (farm: Farm, quoteTokenFarm: Farm, bnbPriceBusd: BigNumber) => {
+  if (farm.quoteToken.symbol === 'BUSD') {
+    return new BigNumber(1)
+  }
+
+  if (farm.quoteToken.symbol === 'wBNB') {
+    return bnbPriceBusd
+  }
+
+  if (!quoteTokenFarm) {
+    return BIG_ZERO
+  }
+
+  if (quoteTokenFarm.quoteToken.symbol === 'wBNB') {
+    return bnbPriceBusd.times(quoteTokenFarm.tokenPriceVsQuote)
+  }
+
+  if (quoteTokenFarm.quoteToken.symbol === 'BUSD') {
+    return new BigNumber(quoteTokenFarm.tokenPriceVsQuote)
+  }
+
+  return BIG_ZERO
+}
+
+const fetchFarmsPrices = async (farms) => {
```

Comment:
I can't type this shit as an array of farms `Farm[]` because of a string / BigNumber typing conflict. The `farms` its receiving are all with serialised BN values, so the types are:

```
tokenAmountMc: string;
quoteTokenAmountMc: string;
tokenAmountTotal: string;
quoteTokenAmountTotal: string;
etc.
```

Whereas in the `Farm` interface, these are all BNs

I'm not 100% why it's happy enough typing the action payload as `Farm[]` here https://github.com/pancakeswap/pancake-frontend/blob/5bfe5f074b493f3d314e310d91bd02186319ad02/src/state/farms/index.ts#L34, which is receiving the same farm object, with strings rather than BNs.

Similarly, I _think_ it's OK typing the argument as `Farm` in the .find/.map functions in the lines below this comment because the values are optional, and it isn't able to infer the types in the farm that's being passed to it.

Annoying, not sure the best solution here, I think the typing of these values as BN may just be wrong?

Reply:
I think the problem is that `Farm` itself is typed improperly, they are all strings (they can't be BigNumber cause they go to Redux), and we convert them to BigNumber in Farms table code.

So I guess the typings could be resolved by changing Farm to string? That might cause some typing issues elsewhere in the code, although I suppose not much (if at all) because improper BigNumber didn't raise any type errors

Reply:
I think you're right Cheems. The issue is that we have some functions expecting these values to be BNs, so those would need to be changed too. I don't think it's too big a job to change these types from BN > string but would ideally do it in another PR IMO.